### ISSUE TITLE
Not properly preserving state of statement with `yield` as part of it

### ISSUE THREAD

**Issue by getify** [facebook/regenerator#206]
The following code does not transform as expected:

```js
var a = 1;
var b = 2;

function *foo() {
  b = a * (yield 2)
}

var it = foo();

it.next();
a = 3;   // mutate `a` while generator is paused
it.next(5);

console.log(a,b);
```

In FF and Chrome, this code produces `3 5`, but in regenerator [and thus babel](http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&playground=false&code=var%20a%20%3D%201%3B%0Avar%20b%20%3D%202%3B%0A%0Afunction%20*foo()%20%7B%0A%20%20b%20%3D%20a%20*%20(yield%202)%0A%7D%0A%0Avar%20it%20%3D%20foo()%3B%0A%0Ait.next()%3B%0Aa%20%3D%203%3B%20%0Ait.next(5)%3B%0A%0Aconsole.log(a%2Cb)%3B%0A), `3 15` is the result.

I believe the issue comes down to:

```js
b = a * (yield 2)
```

In LTR expression parsing, JS evaluates the `a` expression first (while it's still `1`) before `yield`ing `2`. When `5` comes back in, it should be `1 * 5`. However, the transformation has:

```js
context$1$0.t0 = context$1$0.sent;
b = a * context$1$0.t0;
```

The `a` should have had its value captured in the previous state, but instead it's re-evaluated in this state.

Here's a side-by-side screenshot showing the difference:

![screen shot 2015-06-15 at 9 17 30 am](https://cloud.githubusercontent.com/assets/150330/8161881/86b860cc-133f-11e5-86f2-3aae89192d2f.png)

**Comment by benjamn** [comment]
Thanks for the interesting report.

See [this list](https://github.com/facebook/regenerator/blob/37b154955334675318c2346e7e73af460ffce087/lib/emit.js#L989-L991) of reasons why an expression might need to be assigned to a temporary variable before participating in an expression that contains a `yield`. I really don't want to assign all subexpressions to temporary variables, but the general problem of determining if the expression could change value while the generator is yielding seems tricky.

This example makes me think we should pre-evaluate any expression that contains a reference to a variable outside the scope of the generator function, but that's not quite adequate, as you can see in this example:
```js
function *gen() {
  var foo = { bar: 2 };
  return foo.bar * (yield foo);
}
var g = gen();
g.next().bar = 3;
console.log(g.next(5));
// { value: 3 * 5, done: true } in current Regenerator
// { value: 2 * 5, done: true } in Chrome and FF
```
On further reflection, even simple variables are not safe:
```js
function *gen(x) {
  yield (function(y) { x = y });
  return x * (yield 2);
}
var g = gen(2);
g.next()(3);
g.next(); // Same disagreement!
```
In other words, without some really sophisticated analysis of whether an expression is "safe" in the sense that its value provably cannot change while the generator is yielding, I'm not sure there's any reliable way to avoid using temporary variables. What looked like a narrow edge case is actually the common case, for all we can tell.

To be clear, I'm sure I can fix this by using temporary variables more liberally. I just wanted to share my reasoning about why doing anything more intelligent is probably impractical, in case anyone else has any bright ideas.

**Comment by benjamn** [comment]
I also briefly wondered if this behavior was underspecified, leaving room for interpretation, but no: https://people.mozilla.org/~jorendorff/es6-draft.html#sec-multiplicative-operators-runtime-semantics-evaluation

**Comment by getify** [comment]
[Another example](http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&playground=false&code=function%20*foo()%20%7B%0A%20%20var%20x%20%3D%20a.b%20%2B%20(yield)%3B%0A%20%20console.log(x)%3B%0A%7D%0A%0Avar%20a%20%3D%20%7B%0A%20%20%20%20__b%3A0%2C%0A%20%20%20%20get%20b()%7B%20return%20%2B%2Bthis.__b%3B%7D%0A%20%20%7D%2C%0A%20%20it%20%3D%20foo()%3B%0A%0Ait.next()%3B%0Aa.b%3B%0Ait.next(2)%3B) would be if the first expression (`a` in my test case) was a getter (although you may already be thinking about this):

```js
function *foo() {
  var x = a.b + (yield);
  console.log(x);
}

var a = {
    __b:0,
    get b(){ return ++this.__b;}
  },
  it = foo();

it.next();
a.b;
it.next(2);
// should be: 3, regenerator says 4
```

And (again, may already be being covered by your thinking), here's [another illustration](http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&playground=false&code=function%20*foo()%20%7B%0A%20%20%20var%20arr%20%3D%20%5Ba%2Cyield%2Ca%5D%3B%0A%20%20%20console.log(arr)%3B%0A%7D%0A%0Avar%20a%20%3D%201%2C%20it%20%3D%20foo()%3B%0Ait.next()%3B%0Aa%20%3D%203%3B%0Ait.next(2)%3B) of this same sort problem:

```js
function *foo() {
   var arr = [a,yield,a];
   console.log(arr);
}

var a = 1, it = foo();
it.next();
a = 3;
it.next(2);
// should be: [1,2,3], regenerator says [3,2,3]
```

And [yet another](http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&playground=false&code=function%20*foo()%20%7B%0A%20%20a%20%3D%20a.slice(0).concat(yield)%3B%0A%20%20console.log(a)%3B%0A%7D%0A%0Avar%20a%20%3D%20%5B%5D%2C%20it%20%3D%20foo()%3B%0Ait.next()%3B%0Aa.push(1)%3B%0Ait.next(2)%3B):

```js
function *foo() {
  a = a.slice(0).concat(yield);
  console.log(a);
}

var a = [], it = foo();
it.next();
a.push(1);
it.next(2);
// should be: [2], regenerator says [1,2]
```

**Comment by benjamn** [comment]
Not quite fixed because of that last example, but some progress made.

**Comment by getify** [comment]
[Another example](http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&playground=false&code=function%20*foo()%20%7B%0A%20%20var%20x%20%3D%20a%20%2B%20(yield)%3B%0A%20%20console.log(x)%3B%0A%7D%0A%0Avar%20a%20%3D%20%5B1%2C2%5D%2C%20it%20%3D%20foo()%3B%0A%0Ait.next()%3B%0A%0Aa%20%3D%20%5B4%2C5%5D%3B%0A%0Ait.next(%22%2C3%22)), not sure if this is already covered by your previous fixes:

```js
function *foo() {
  var x = a + (yield);
  console.log(x);
}

var a = [1,2], it = foo();

it.next();

a = [4,5];

it.next(",3");
// should be: "1,2,3", regenerator says "4,5,3"
```

**Comment by getify** [comment]
BTW, [here's an example](https://gist.github.com/getify/c8e5639f322d00cfe244) I was trying to explore if there's any way to get these sequencing side-effects even with a literal. I don't think this will work, but I figured I'd just list it here for posterity, and in case it jogs any other thoughts of other ways this could be "hacked".

**Comment by benjamn** [comment]
Thanks for all the examples, @getify! I've added them as tests and pushed a fix that makes them all pass. The trickiest one was the `a.slice(0).concat(yield)`, because it requires transforming to something like
```js
var t0 = a.slice(0);
var result = t0.concat.call(t0, yield);
```
Unless someone can demonstrate that literal expressions are volatile too, I think this can be closed.

**Comment by getify** [comment]
<3 thanks for the quick fixes. will let my brain keep churning on any literal volatility and report back if I come up with anything. :)

**Comment by benjamn** [comment]
I'm waiting for Travis tests to pass before I publish to NPM, but I've update the sandbox so you can play with examples there if it helps: http://facebook.github.io/regenerator/

**Comment by getify** [comment]
Hmmm... this test case fails to compile in that sandbox:

```js
function *foo() {
  var x = a + (yield);
  console.log(x);
}

var a = [1,2], it = foo();

it.next();

a = [4,5];

it.next(",3");
```

It complains apparently about the `a + (yield)`. If I change that to `a + (yield 1)` it works fine. Is this a separate bug, or a regression?

**Comment by benjamn** [comment]
Ah yes, I'm pretty sure that's an `esprima-fb` bug that would be fixed by using Acorn, which is something I've been meaning to do.
